### PR TITLE
HTTP2 improvements

### PR DESCRIPTION
**Pull request by adam-fowler**
- Added `HTTP2ServerConnectionManager` to manage graceful close and idle, close, maxAge timeouts. This is heavily based on the gRPC ServerConnectionManager. 
- Added `HTTP2StreamChannel` to serve a single HTTP request/response pair instead of using the `HTTP1Channel`
- Moved HTTP2 tests into their own target

### CODE REVIEW

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 1]
```diff
@@ -16,39 +16,56 @@ import HTTPTypes
...
-        let negotiatedHTTPVersion: EventLoopFuture<NIONegotiatedHTTPVersion<HTTP1Channel.Value, (NIOAsyncChannel<HTTP2Frame, HTTP2Frame>, NIOHTTP2Handler.AsyncStreamMultiplexer<HTTP1Channel.Value>)>>
+        let negotiatedHTTPVersion: EventLoopFuture<NIONegotiatedHTTPVersion<HTTP1ConnectionOutput, HTTP2ConnectionOutput>>
         public let channel: Channel
     }
 
     /// HTTP2 Upgrade configuration
     public struct Configuration: Sendable {
+        /// Idle timeout, how long connection is kept idle before closing
+        public var idleTimeout: Duration?
+        /// Maximum amount of time to wait before all streams are closed after second GOAWAY
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:41]: to wait on what?

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 2]
```diff
@@ -16,39 +16,56 @@ import HTTPTypes
...
+        /// Maximum amount of time a connection can be open
+        public var maxAgeTimeout: Duration?
         /// Configuration applieds to HTTP2 stream channels
         public var streamConfiguration: HTTP1Channel.Configuration
 
         ///  Initialize HTTP2UpgradeChannel.Configuration
         /// - Parameters:
-        ///   - additionalChannelHandlers: Additional channel handlers to add to HTTP2 connection channel
+        ///   - idleTimeout: How long connection is kept idle before closing
+        ///   - maxGraceCloseTimeout: Maximum amount of time to wait before all streams are closed after second GOAWAY
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:51]: Likewise, what's it waiting on?

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 3]
```diff
@@ -16,39 +16,56 @@ import HTTPTypes
...
+            self.gracefulCloseTimeout = gracefulCloseTimeout
             self.streamConfiguration = streamConfiguration
         }
     }
 
     private let sslContext: NIOSSLContext
     private let http1: HTTP1Channel
-    public var responder: HTTPChannelHandler.Responder { self.http1.responder }
+    private let http2Stream: HTTP2StreamChannel
+    let configuration: Configuration
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:68]: Does this configuration become `public`?

File: Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                return .none
+
+            case .closed:
+                return .none
+            }
+        }
+
+        enum ReceivedPingResult {
+            case sendPingAck(pingData: HTTP2PingData)
+            case enhanceYouCalmAndClose(lastStreamId: HTTP2StreamID)
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift:97]: What does this mean?

File: Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+//===----------------------------------------------------------------------===//
+
+import HTTPTypes
+import HummingbirdCore
+import Logging
+import NIOCore
+import NIOHTTPTypes
+import NIOHTTPTypesHTTP2
+
+/// Child channel for processing HTTP1
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift:22]:
```suggestion
/// Child channel for processing HTTP2
```

File: Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+struct HTTP2StreamChannel: ServerChildChannel, HTTPChannelHandler {
+    typealias Value = NIOAsyncChannel<HTTPRequestPart, HTTPResponsePart>
+    typealias Configuration = HTTP1Channel.Configuration
+
+    ///  Initialize HTTP1Channel
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift:27]:
```suggestion
    ///  Initialize HTTP2Channel
```

File: Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    /// - Parameters:
+    ///   - responder: Function returning a HTTP response for a HTTP request
+    ///   - configuration: HTTP1 channel configuration
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift:30]:
```suggestion
    ///   - configuration: HTTP2 channel configuration
```

File: Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    init(
+        responder: @escaping HTTPChannelHandler.Responder,
+        configuration: Configuration = .init()
+    ) {
+        self.configuration = configuration
+        self.responder = responder
+    }
+
+    /// Setup child channel for HTTP1
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2StreamChannel.swift:39]:
```suggestion
    /// Setup child channel for HTTP2
```

**Review by Joannis — CHANGES_REQUESTED** [review]
Initial review, I didn't have a complete look through yet

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2Channel.swift:41]: for client response

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift (hunk 2 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2Channel.swift:51]: for client response

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2Channel.swift:68]: For consistency with other channels it should be public

File: Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift]: Enhance you calm is an unofficial status code Twitter defined for when a client sends too many requests (In this case too many pings). I've added a comment

Comment (dannflor) [reply Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift]: I believe twitter actually spells it "Enhance your calm", not "Enhance you calm"

Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift]: 🤣 that's a typo

**Review by Joannis — APPROVED** [review]

**Pull request merged**